lib/remap-lang/src/parser.rs

```diff
@@ -301,65 +353,103 @@ impl Parser<'_> {
-        let mut segments = vec![];
-        for segment in pair.into_inner() {
-            match segment.as_rule() {
-                R::path_field => segments.push(self.path_field_from_pair(segment)?),
-                R::path_coalesce => segments.append(&mut self.path_coalesce_from_pair(segment)?),
-                R::path_index => segments
-                    .last_mut()
-                    .get_or_insert(&mut "".to_owned())
-                    .push_str(segment.as_str()),
-                _ => todo!(),
-            }
+    fn path_segment_from_pair(&self, pair: Pair<R>) -> Result<path::Segment> {
+        let segment = pair.into_inner().next().ok_or(e(R::path_segment))?;
+
+        match segment.as_rule() {
+            R::path_field => self.path_field_segment_from_pair(segment),
+            R::path_coalesce => self.path_coalesce_segment_from_pair(segment),
+            _ => Err(e(R::path_segment)),
         }
+    }
 
-        Ok(segments)
+    fn path_field_segment_from_pair(&self, pair: Pair<R>) -> Result<path::Segment> {
+        self.path_field_from_pair(pair).map(path::Segment::Field)
     }
 
-    fn path_field_from_pair(&self, pair: Pair<R>) -> Result<String> {
+    fn path_coalesce_segment_from_pair(&self, pair: Pair<R>) -> Result<path::Segment> {
+        let fields = pair
+            .into_inner()
+            .map(|pair| self.path_field_from_pair(pair))
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(path::Segment::Coalesce(fields))
+    }
+
+    fn path_field_from_pair(&self, pair: Pair<R>) -> Result<path::Field> {
         let field = pair.into_inner().next().ok_or(e(Rule::path_field))?;
 
         match field.as_rule() {
             R::string => {
                 let string = field.into_inner().next().ok_or(e(R::string))?;
-                self.escaped_string_from_pair(string)
+                Ok(path::Field::Quoted(self.escaped_string_from_pair(string)?))
             }
-            R::ident => Ok(field.as_str().to_owned()),
-            _ => Err(e(Rule::path_field)),
+            R::ident => Ok(path::Field::Regular(field.as_str().to_owned())),
+            _ => Err(e(R::path_field)),
         }
     }
 
-    fn path_coalesce_from_pair(&self, pair: Pair<R>) -> Result<Vec<String>> {
-        pair.into_inner()
-            .map(|pair| self.path_field_from_pair(pair))
-            .collect::<Result<_>>()
+    fn path_index_from_pair(&self, pair: Pair<R>) -> Result<path::Segment> {
+        let index = pair
+            .into_inner()
+            .next()
+            .ok_or(e(R::path_index))?
+            .as_str()
+            .parse::<usize>()
+            .map_err(|_| e(R::path_index_inner))?;
+
+        Ok(path::Segment::Index(index))
     }
 
     /// Parse a [`Variable`] value, e.g. "$foo"
-    fn variable_from_pair(&self, pair: Pair<R>) -> Result<Expr> {
-        let ident = pair.into_inner().next().ok_or(e(R::variable))?;
+    fn variable_from_pair(&self, pair: Pair<R>) -> Result<Variable> {
```

Comment:
I'm not sure I understand why this needs to work differently from `path_from_pair`?

Is the plan to tidy this up with #5350?

Reply:
It differs exactly because of the reasons listed in the issue you linked.

That is to say, a variable right not _has_ to include a top-level identifier: `$foo`, you can't write `$`, whereas `.` is a valid path.

So, this means we have to parse paths attached to a variable a bit differently from paths.

It would be cleaned up if we did what I described in #5350, yes, but I we'll have to reach consensus first.

Reply:
So, they are both essentially just a list of segments, but they differ in that one requires at least one `path_segment` segment, and the other is ok to have none?

Couldn't the code to parse a `Vec` of segments be extracted out and shared between both?

Pest code is not the easiest to follow, so there's a good chance I'm missing something! But I think it is important to aim for as much simplicity here as possible.

Reply:
This is related to my comment here: https://github.com/timberio/vector/pull/5277#discussion_r532074268

I'll probably do another pass in a follow-up PR, but right now these two differ too much to unify more without changing the grammar (or without making the parser even more complex).

It all stems from these two rules that differ:

```
variable = ${ "$" ~ ident ~ path_index* ~ ("." ~ path_segments)?  }
```

```
path             = ${ "." ~ path_segments? ~ !(WHITESPACE* ~ ("." | "\"")) }
path_segments    = ${
    (
        ((path_segment | path_index+) ~ path_index*)
        ~ ("." ~ path_segment ~ path_index*)*
    )
    | path_index+
}
path_segment     = ${ path_field | path_coalesce }
```

Reply:
Got it.

At the risk of over labouring a point, is there a reason they couldn't share the rules:

```
path             = ${ ("." | "$") ~ path_segments? ~ !(WHITESPACE* ~ ("." | "$" | "\"")) }
path_segments    = ${
    (
        ((path_segment | path_index+) ~ path_index*)
        ~ (("." | "$") ~ path_segment ~ path_index*)*
    )
    | path_index+
}
path_segment     = ${ path_field | path_coalesce }
```

Validation dissallowing a lone `$` could still occur later on in code?

Yeah, I'm all for #5350.

Reply:
As long as we don't have #5350 yet, I like to keep a conceptual difference between paths and variables. If we do the check later in the code, then we lose the useful error message we get from Pest.

We do still want to improve error messages in general, but we haven't done so yet, unfortunately.

Reply:
Yes, good point on the error messages.